### ISSUE TITLE
dns masq should have resources defined

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#29055]
kubedns has resource/limits but not the dnsmasq cache.

**Comment by girishkalele** [comment]
100m is probably enough for CPU.
For memory resources, need to fill up dnsmasq cache with its default cache-limit using the dnsbench sample queries file and measure memory utilization.

**Comment by girishkalele** [comment]
This Google 20% project looks like another DNS benchmark datapoint we could use.

https://code.google.com/archive/p/namebench/

**Comment by bowei** [comment]
# Dnsmasq memory consumption

To reproduce these results:

https://github.com/bowei/contrib/tree/dnsmasq-mem/dnsmasq-mem

## TLDR

dnsmasq in its current configuration caps the size of its cache to 10k records.
(See `src/options.c`) It will not use more than 3mb for typical workloads. When
the cache limit was removed, we get the following memory usage scaling:

cache size | memory footprint
-----------|-----------------
10k        | 3 mb
100k       | 13 mb
1M         | 112 mb

# Dnsmasq notes

* Does not cache entry types that are unbounded in memory usage (e.g.
  TXT records).
* Each record in a response consumes a single entry in the cache. E.g. if 
  foo.com has 5 A records returned, then this will consume 5 entries.
* If a response with multiple records does not fit in the cache, it is not
  cached.

# Evaluation details

Evaluation consists of a bind9 server configured to result synthetic records
(A, AAAA, TXT, SRV), a dnsmasq instance and dnsperf client:

````
  +------+       +---------+       +---------+
  | bind | <---- | dnsmasq | <---- | dnsperf |
  +------+       +---------+       +---------+
````

`make` will build the docker images and run the memory consumption harness. The
resulting `run.log` will contain the resulting memory used by dnsmasq after
fully populating its cache.

**Comment by bprashanth** [comment]
Makes sense, so we need to define resource request on https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.in#L85. are you suggesting 100m cpu and the default 10k/10mb memory? if cpu is left unspecified I believe it gets the lowest possible allocation under load (2 shares).

**Comment by bprashanth** [comment]
@bowei Hmm, you are a member of the Kubernetes org but I can't assign the bug to you. I think you got unassigned on my last comment, perhaps because you jumped in/out of membership somehow?
